ethcore/engines/authority-round/src/lib.rs

```diff
@@ -967,35 +977,29 @@ impl AuthorityRound {
 		});
 	}
 
-	fn handle_empty_step_message(&self, empty_step: EmptyStep) {
+	fn store_empty_step(&self, empty_step: EmptyStep) {
 		self.empty_steps.lock().insert(empty_step);
 	}
 
-	fn generate_empty_step(&self, parent_hash: &H256) {
+	/// Build an EmptyStep and broadcast it to the network.
+	fn emit_empty_step(&self, parent_hash: &H256) {
 		let step = self.step.inner.load();
 		let empty_step_rlp = empty_step_rlp(step, parent_hash);
 
 		if let Ok(signature) = self.sign(keccak(&empty_step_rlp)).map(Into::into) {
-			let message_rlp = empty_step_full_rlp(&signature, &empty_step_rlp);
-
 			let parent_hash = *parent_hash;
 			let empty_step = EmptyStep { signature, step, parent_hash };
 
 			trace!(target: "engine", "broadcasting empty step message: {:?}", empty_step);
-			self.broadcast_message(message_rlp);
-			self.handle_empty_step_message(empty_step);
-
+			if let Ok(c) = self.upgrade_client_or("could not broadcast empty step message") {
+				self.store_empty_step(empty_step);
+				c.broadcast_consensus_message(empty_step_full_rlp(&signature, &empty_step_rlp));
```

Comment:
I don't know if it makes any difference, but previously a message was broadcasted before storing it and irregardless of whether client upgrade ref was successfull

Reply:
Indeed, I think that behaviour is slightly suspicious. I have no proof that it was wrong, but I think the new behaviour is safer.